### PR TITLE
(PDB-2286) Add ability to always store catalog jsonb fields, (PDB-2290) Munge edges and resources jsonb for storage

### PR DESCRIPTION
**Pull request by ajroetker**
This PR fixes a bug with historical catalogs where a user could turn
off historical catalogs and break some of the PE views because the JSONB
column storage was also turned off. This new atom will always be set to
true in PE (as opposed to the historical-catalogs' atom which is config
dependent).

This PR aligns the edges and resources jsonb format with the return
format from querying the catalogs endpoint, this means we do not have to
parse and munge jsonb in export, the new endpoints and sync.

### CODE REVIEW

File: src/puppetlabs/puppetdb/scf/storage.clj [hunk 1]
```diff
@@ -351,16 +351,32 @@
    (comp first sql/result-set-seq)))
 
 (def store-catalogs-historically? (atom false))
+(def store-catalogs-jsonb-columns? (atom false))
```
Comment (mullr) [inline src/puppetlabs/puppetdb/scf/storage.clj]: Can you put a comment here saying the same thing you did in your commit message about how these two are different?

**Comment by ajroetker** [comment]
@mullr added the comments suggested by your comments :boom:

File: src/puppetlabs/puppetdb/export.clj [hunk 2]
```diff
@@ -92,8 +91,14 @@
...
                                            (maybe-anonymize anonymize-fn anon-config)
                                            (export-data->tar-items entity)
-                                           (add-tar-entries tar-writer)))]]
-      (query-fn query-api-version ["from" entity] nil query-callback-fn))))
+                                           (add-tar-entries tar-writer)))
+                  entity* (if (and (= entity "catalogs")
+                                   @storage/store-catalogs-jsonb-columns?)
+                            ;; *Warning* this can only be used in PE so it
+                            ;; *cannot be tested against in the FOSS repo
+                            "historical_catalogs"
```
Comment (senior) [inline src/puppetlabs/puppetdb/export.clj]: I guess what this needs is a plug point for PE?

File: src/puppetlabs/puppetdb/scf/storage.clj [hunk 3]
```diff
@@ -350,17 +350,43 @@
     certname]
    (comp first sql/result-set-seq)))
 
+;; `store-catalogs-historically?` is used for toggling historical catalog
+;; storage, the ability to toggle this variable is in the PE extensions repo.
```
Comment (senior) [inline src/puppetlabs/puppetdb/scf/storage.clj]: Seems like we can say this is PE only and configurable and leave out repo information

**Comment by senior** [comment]
:+1: minor comment on a comment

File: src/puppetlabs/puppetdb/export.clj (hunk 2 above)
Comment (ajroetker) [inline src/puppetlabs/puppetdb/export.clj]: Yeah I don't exactly what we'd want to do here, maybe make an atom that PE can use to change the query entities or such eventually.

File: src/puppetlabs/puppetdb/scf/storage.clj (hunk 3 above)
Comment (ajroetker) [inline src/puppetlabs/puppetdb/scf/storage.clj]: :+1:

**Comment by ajroetker** [comment]
@senior fixed the comment!

File: src/puppetlabs/puppetdb/catalogs.clj [hunk 4]
```diff
@@ -158,7 +158,9 @@
    :file (s/maybe s/Str)
    :line (s/maybe s/Int)
    :parameters {s/Any s/Any}
-   :resource s/Str
+   ;; On historical-catalogs this is an optional-key but this is required on
+   ;; "normal" catalogs
+   (s/optional-key :resource) s/Str
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/catalogs.clj]: I'd rather this be addressed by us storing resource hashes. I think you could find uses for them and I think not having them introduces unneeded inconsistency between the historical and present day views.

Comment (ajroetker) [inline src/puppetlabs/puppetdb/catalogs.clj]: Either way I'd like to address it in another patch, it requires some nontrivial changes to the storage code.

File: test/puppetlabs/puppetdb/acceptance/cli.clj [hunk 5]
```diff
@@ -49,7 +49,7 @@
 
       (svc-utils/call-with-single-quiet-pdb-instance
        (let [anon-out-map (tar->map anon-out-file)
-             anon-certname (-> anon-out-map (get "reports") first val (get "certname"))]
+             anon-certname (some-> anon-out-map (get "reports") first val (get "certname"))]
```
Comment (wkalt) [inline test/puppetlabs/puppetdb/acceptance/cli.clj]: why is this needed?

File: src/puppetlabs/puppetdb/export.clj (hunk 2 above)
Comment (wkalt) [inline src/puppetlabs/puppetdb/export.clj]: will the export that results from this be importable?

Comment (wkalt) [inline src/puppetlabs/puppetdb/export.clj]: ah got it.  they don't overwrite in PE
